docs/dev/relay_op_strategy.rst

```diff
@@ -0,0 +1,256 @@
+            wrap_topi_schedule(my_op_schedule_lib),
+            name="my_implementation_lib",
+            plevel=20)
+
+
+Further, we can add implementation specialized for a certain range of shapes.
+The code below shows an example of dense strategy that adds an implementation
+that is specialized for ``m`` greater than 16. The main difference between
+hardcode python condition like examples above and specialized condition is that
+it allows TVM to generate multiple kernels when the input tensors have symbolic
+shapes. The compile engine will generate a dispatch function that invokes the
+specialized kernel first when the corresponding condition is met; otherwise,
+invoke the kernel that has no associated specialized condition (``dense_common``
+in this example). This part is still work in progress. More details will be
+provided after it is done.
+
+.. code:: python
+
+    def dense_strategy(attrs, inputs, out_type, target):
+        m = inputs[0].shape[0]
+        strategy = _op.OpStrategy()
+        strategy.add_implementation(
+            wrap_compute_dense(dense_compute1),
+            wrap_topi_schedule(dense_schedule1),
+            name="dense_common")
+
+        with tvm.te.SpecializedCondition(m > 16):
+            strategy.add_implementation(
+                wrap_compute_dense(dense_compute2),
+                wrap_topi_schedule(dense_schedule2),
+                name="dense_for_large_m",
+                plevel=15)
+
+        return strategy
+
+
+Register strategy for a new target
+----------------------------------
+
+There are two ways to register strategies for a new target. The more
+straightforward one is adding a new target file in the directory
+``python/tvm/relay/op/strategy``. You only need to customize the strategy for
+ops that have been implemented for this new target and reuse the generic
+strategies for the rest.
+
+Alternatively, you can also register the strategy for the new target outside the
+TVM python library. The following code snippet shows an example how to do
+so. You can find more examples in ``vta/python/vta/top/op.py``.
+
+.. code:: python
+
+    @relay.op.strategy.someop_strategy.register("mytarget")
+    def someop_strategy_mytarget(attrs, inputs, out_type, target):
+        ...
+
+
+Select implementation from op strategy
+--------------------------------------
+
+During the compilation, Relay compile engine needs to determine which
+implementation to use for an operator when there are multiple. The selection
+policy works as follows.
+
+When the input tensors to an operator or a fused op all have constant shapes,
+the compile engine first finds the best implementation based on AutoTVM tuning
+logs. If there is no implementation that is an AutoTVM template or all AutoTVM
+templates have fallback configs, the implementation with highest priority level
+will then be chosen. Implementations with same priority level in this case leads
+to an undefined behavior, and any of them might be selected.
```

Comment:
this is not related to this PR, but shouldn't this undefined behavior trigger an error?

Reply:
I think it's fine tho. Throwing an error seems too serious to me (and users won't know what does that mean, because the error is for developers). If a developer assigns the same priority level to two implementations, op strategy selects either the one with better tuning result, or the one shows up earlier. It might be fine if we directly say "the first added implementation will be selected for the same priority level", but I am also fine with the current description since it implies we do not recommend developers to use the same priority level at all.

Reply:
Agree with @comaniac. It might be too strict to trigger an error just because two implementation has the same plevel.

Reply:
Thanks, that makes sense